### PR TITLE
Initial etcd removal controller implementation

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

- Initial project structure
- Implement the etcd removal controller to remove etcd members upon their nodes being deleted from the cluster

#### What issue(s) does this fix?
* Fixes https://github.com/containership/cloud.provision/issues/402

#### Additional Considerations

- This controller must run on a master so it can mount etcd certs, which means even more load on masters. This could be avoided by putting the necessary certs in secrets. For now, this solution seems more appropriate to me because it aligns with our existing security model.
- We could probably set the sync interval higher to avoid extraneous checks while the node still exists in cloud but is in the process of being deleted. I figured this is fine for now and we can optimize later if we see fit.

#### Setup

- Provision a Kubernetes v1.12.1 cluster on DO with 6 masters and 2 workers
- Edit development deployment yaml included in this PR to include correct etcd endpoint (make sure to include the port, `:2379`)

#### Instructions

- Deploy development manifest
- Scale down master node pool from 6 to 3 through UI
- See that correct etcd members are removed, as expected:

```
kubectl -n containership-core logs -f infrastructure-controller-59bb46c5b5-lvkzp
```
```
2018-11-20T17:35:06.240Z        INFO    Starting Containership Infrastructure Controller...
2018-11-20T17:35:06.241Z        INFO    Version: unknown (d3e2746) built @ 2018-11-20 17:32:31 +0000 UTC
2018-11-20T17:35:06.241Z        INFO    Go Version: go1.10.1
...
2018-11-20T17:35:06.242Z        INFO    Using in cluster k8s config
2018-11-20T17:35:06.254Z        INFO    Setting up event handlers
2018-11-20T17:35:06.255Z        INFO    Starting controller
2018-11-20T17:35:06.356Z        INFO    Starting workers
2018-11-20T17:35:06.356Z        INFO    Started workers
2018-11-20T17:36:19.525Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:19.525Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:20.016Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:20.158Z        INFO    Missing Kubernetes node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" exists in cloud, will skip etcd member remove
2018-11-20T17:36:20.207Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:20.207Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:20.339Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:20.474Z        INFO    Missing Kubernetes node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" exists in cloud, will skip etcd member remove
2018-11-20T17:36:21.160Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:21.160Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:21.304Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:21.433Z        INFO    Missing Kubernetes node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" exists in cloud, will skip etcd member remove
2018-11-20T17:36:21.567Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:21.567Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:21.712Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:21.947Z        INFO    Missing Kubernetes node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" exists in cloud, will skip etcd member remove
2018-11-20T17:36:22.022Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:22.022Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:22.201Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:22.339Z        INFO    Missing Kubernetes node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" exists in cloud, will skip etcd member remove
2018-11-20T17:36:22.381Z        INFO    Found etcd member named "a3feb4a7-17b4-475d-844c-6fd1d06aa646" with no corresponding Kubernetes node
2018-11-20T17:36:22.385Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:22.520Z        DEBUG   Checking for existence of node "a3feb4a7-17b4-475d-844c-6fd1d06aa646" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:22.654Z        INFO    Missing Kubernetes node "a3feb4a7-17b4-475d-844c-6fd1d06aa646" exists in cloud, will skip etcd member remove
2018-11-20T17:36:24.152Z        INFO    Found etcd member named "a3feb4a7-17b4-475d-844c-6fd1d06aa646" with no corresponding Kubernetes node
2018-11-20T17:36:24.157Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:24.321Z        DEBUG   Checking for existence of node "a3feb4a7-17b4-475d-844c-6fd1d06aa646" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:24.456Z        INFO    Missing Kubernetes node "a3feb4a7-17b4-475d-844c-6fd1d06aa646" exists in cloud, will skip etcd member remove
2018-11-20T17:36:26.458Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:26.463Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:26.600Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:26.836Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:30.176Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:30.180Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:30.368Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:30.518Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:31.213Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:31.214Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:31.400Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:31.553Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:31.605Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:31.605Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:31.753Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:31.888Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:31.925Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:31.931Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:32.167Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:32.293Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:34.206Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:34.206Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:34.340Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:34.468Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:36.394Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:36.394Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:36.527Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:36.654Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:36.709Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:36.710Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:36.845Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:37.000Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:37.165Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:37.165Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:37.295Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:37.424Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:37.475Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:37.475Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:37.610Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:37.742Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:37.776Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:37.776Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:37.912Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:38.044Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:40.229Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:40.230Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:40.361Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:40.486Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:41.231Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:41.236Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:41.369Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:41.496Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:41.647Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:41.648Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:41.782Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:41.906Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:41.933Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:41.933Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:42.065Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:42.191Z        INFO    Missing Kubernetes node "325ec4fc-29e9-4e3e-a487-1e4b373599df" exists in cloud, will skip etcd member remove
2018-11-20T17:36:44.197Z        INFO    Found etcd member named "325ec4fc-29e9-4e3e-a487-1e4b373599df" with no corresponding Kubernetes node
2018-11-20T17:36:44.198Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:44.331Z        DEBUG   Checking for existence of node "325ec4fc-29e9-4e3e-a487-1e4b373599df" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:44.439Z        INFO    Requesting etcd member remove for member with name "325ec4fc-29e9-4e3e-a487-1e4b373599df"
2018-11-20T17:36:44.464Z        INFO    etcd member "325ec4fc-29e9-4e3e-a487-1e4b373599df" (ID 117b79fca0fa94a3) removed successfully
2018-11-20T17:36:50.252Z        INFO    Found etcd member named "a3feb4a7-17b4-475d-844c-6fd1d06aa646" with no corresponding Kubernetes node
2018-11-20T17:36:50.257Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:50.395Z        DEBUG   Checking for existence of node "a3feb4a7-17b4-475d-844c-6fd1d06aa646" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:50.503Z        INFO    Requesting etcd member remove for member with name "a3feb4a7-17b4-475d-844c-6fd1d06aa646"
2018-11-20T17:36:50.522Z        INFO    etcd member "a3feb4a7-17b4-475d-844c-6fd1d06aa646" (ID 1709e3402becb816) removed successfully
2018-11-20T17:36:51.673Z        INFO    Found etcd member named "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" with no corresponding Kubernetes node
2018-11-20T17:36:51.673Z        DEBUG   Getting all etcd node pools
2018-11-20T17:36:51.807Z        DEBUG   Checking for existence of node "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" in etcd node pool "cf344615-897a-471d-bb6f-241e551540be"
2018-11-20T17:36:51.918Z        INFO    Requesting etcd member remove for member with name "bd576a72-7368-4cf3-9122-bf27aeaa5e5e"
2018-11-20T17:36:51.935Z        INFO    etcd member "bd576a72-7368-4cf3-9122-bf27aeaa5e5e" (ID 9b4014ad4aadc817) removed successfully
```

```
ETCDCTL_API=3 etcdctl --endpoints="etcd.3cc9c20a-8d6c-47ed-bc73-38ee8786c6b3.62e4e86f-fe2e-4740-a814-a950bf377daf.stage.cloud.cship.io:2379"     --cacert=/etc/kubernetes/pki/etcd/ca.crt --cert=
```
```
/etc/kubernetes/pki/etcd/healthcheck-client.crt --key=/etc/kubernetes/pki/etcd/healthcheck-client.key member list
61db781c0f0b5e4, started, 17083c80-21fb-4c40-b295-c6feb52e626a, https://10.131.95.160:2380, https://10.131.95.160:2379
4db19e6da04dce8f, started, 0896f14e-b4a0-4d50-bd6a-40bc6d92c4ed, https://10.131.95.158:2380, https://10.131.95.158:2379
ef87614bde5135ee, started, f1a8d400-93ce-4364-938e-7ee51e2709b6, https://10.131.96.66:2380, https://10.131.96.66:2379
```

### Dependencies
* None (but a `cloud.api` PR is needed for integration)

### CODE REVIEW

**Comment by mattkelly** [comment]
Marked as WIP primarily due to need for:

- Unit testing
- CI

**Comment by mattkelly** [comment]
Added some minimal tests for now. The bulk of the required unit testing requires a mock of the cloud client, which is not yet available :/

**Comment by mattkelly** [comment]
Removing `status/wip`; I think this initial implementation is good for final review.

File: pkg/controller/etcd_removal.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	"github.com/containership/infrastructure-controller/pkg/env"
+	"github.com/containership/infrastructure-controller/pkg/etcd"
+	"github.com/containership/infrastructure-controller/pkg/node"
+)
+
+const (
+	delayBetweenRequeues = 30 * time.Second
+
+	// Don't requeue in order to avoid excessive requests to cloud for things
+	// that we're going to naturally retry on the sync interval anyway
```
Comment (ashleyschuett) [inline pkg/controller/etcd_removal.go]: 👍

File: deploy/development/infrastructure-controller.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    spec:
+      serviceAccountName: containership-admin
+      tolerations:
+      - effect: NoSchedule
+        key: node-role.kubernetes.io/master
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      containers:
+      - name: infrastructure-controller
+        image: containership/infrastructure-controller
```
Comment (ashleyschuett) [inline deploy/development/infrastructure-controller.yaml]: should there be a version associated with that? or wait until we add a tag?

Comment (mattkelly) [reply deploy/development/infrastructure-controller.yaml]: Manifests here are just for local dev, so using `latest` is fine IMO

File: pkg/controller/etcd_removal.go (hunk 1 above)
Comment (mattkelly) [reply pkg/controller/etcd_removal.go]: 🤘

File: pkg/node/node.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+			etcdPools = append(etcdPools, np)
+		}
+	}
+
+	return etcdPools, nil
+}
+
+// nodeExistsInPool returns true if the node with the given ID exists in the
+// given node pool, else false.
+// Because a node pool object does not contain the actual node IDs, additional
```
[the remark this answers is not in the document]
Comment (mattkelly) [reply pkg/node/node.go]: I can see that. Updated.

**Comment by mattkelly** [comment]
Updated and filed https://github.com/containership/csctl/issues/57 for implementing a cloud client mock so that we can add better tests here in the (hopefully near) future.

File: cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/containership/infrastructure-controller/pkg/env"
+)
+
+func main() {
+	log.Info("Starting Containership Infrastructure Controller...")
+	log.Infof("Version: %s", buildinfo.String())
+	log.Infof("Go Version: %s", runtime.Version())
+
+	if err := env.Init(); err != nil {
+		log.Fatalf("Error initialization environment: %v", err)
```
Comment (mattkelly) [inline cmd/main.go]: 🤦‍♂️ I shall live in eternal shame

**Pull request merged**